### PR TITLE
add terminal widget [DRAFT]

### LINKED ISSUE #8: Feature: virtual terminal
## Feature

add a virtual terminal widget, so we can show the live stdout and stderr of the `run` function

### Is your feature request related to a problem? Please describe

problem: when running the app without a terminal, then stdout/stderr are not visible

### Describe the solution you'd like

challenges:

* capture stdout and stderr &rarr; function call to `run_function` or `subprocess`
* correctly interleave stdout and stderr
* render terminal control commands &rarr; virtual terminal (tty) / pseudo terminal (pty)
  * https://github.com/Saadmairaj/tkterminal
  * https://stackoverflow.com/questions/10239965/giving-a-command-in-a-embedded-terminal
  * https://stackoverflow.com/questions/7253448/how-to-embed-a-terminal-in-a-tkinter-application
  * https://stackoverflow.com/questions/62335989/redirect-terminal-output-to-tkinter

example: capture stdout with `io.TextIOBase`
https://stackoverflow.com/questions/51527750/capture-print-output-from-multiple-threads-in-python

<details>

```py
import threading
import time
import sys
import io

class demo_copy_stdout():

  def __init__(self):

    def run_function():
      for i in range(0, 100):
        print(*sys.argv, i)
        time.sleep(1)

    self.run_function = run_function

    def main_function_wrapper(*argv):
        sys.argv = argv # set arguments
        # note: run_function takes no arguments
        return self.run_function()

    print("main_thread create")
    self.main_thread = threading.Thread(target=main_function_wrapper, args=("toy.py", "hello",))

    class CopyStdout(io.TextIOBase):
        def __init__(self, target):
            self.target = target
            #self.stringio = io.StringIO()
        def write(self, s):
            writecount = self.target.write(s)
            written = s[:writecount]
            #self.stringio.write(written)
            # TODO send to GUI
            with open("stdout.txt", "a") as f:
                f.write(written)
            return writecount
    print("copying stdout to tee.txt")
    original_stdout = sys.stdout
    sys.stdout = CopyStdout(sys.stdout)

    print("main_thread start")
    self.main_thread.start() # run

    if True:
        print("main_thread wait ...")
        self.main_thread.join() # wait
        print("main_thread done")
        sys.stdout = original_stdout # stop copying stdout

x = demo_copy_stdout()
```

</details>

### PR DESCRIPTION
**Pull request by milahu**
### What is the purpose of this pull request? (put an "X" next to item)

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Other

### What changes did you make? (Give an overview)

add a fully-featured bash terminal to the GUI
to show the output of the process
and to allow the user to "play" with a bash terminal

### Which issue (if any) does this pull request address?
Fixes #8

### CODE REVIEW

**Comment by FredHappyface** [comment]
Thanks again for the work on this!

Just taken a quick look (aware that is is WIP so feel free to take these comments with a pinch of salt)

One thing that caught my eye was the inclusion of a local pysimplegui. If you've added/ changed functionality then I'd suggest raising that upstream as it may benefit other users too. And if you don't fancy that then can you monkey patch it or add the functionality in a different file (may be worth vendoring depending on the size etc)

Just a forewarning that I won't be accepting a pr with +20k lines as I don't have the time to review that

**Comment by milahu** [comment]
basically working on linux, tested with [autosub](https://github.com/abhirooptalasila/AutoSub)

just needed some glue code for the file inputs
cos autosub has `type=str`
but cli2gui needs `type=argparse.FileType("rb")`

<details>
<summary>
glue code
</summary>

```py
def main():
    # run the CLI version
    parser = argparse.ArgumentParser(description="AutoSub")
    parser.add_argument(
        "--file",
        type=argparse.FileType("rb"), # -> value is io.BufferedReader(name="/path/to/file", ...)
        required=False,
        help="Input video file"
    )
    # ...
    args = parser.parse_args()
    run(args)

def run(args):
    # glue code between cli2gui and main
    # convert io.BufferedReader to str
    for file_key in ["file", "model", "scorer"]:
        file = getattr(args, file_key)
        if file:
            name = file.name
            file.close()
            setattr(args, file_key, name)
    # process args.file ...

cli2gui_decorate = Cli2Gui(
    run_function=run,
    auto_enable=True,
    program_name="AutoSub",
    program_description="Generate subtitles for videos using offline speech recognition",
)

# The gui function can be used as a GUI entrypoint
# Example: python -m mymodule:main.gui
def gui():
    if len(sys.argv) > 1: # TODO better
        main()
    else:
        gui_main = cli2gui_decorate(main)
        gui_main()
```

</details>

todo: make it work on windows, using [busybox-w32](https://github.com/rmyorston/busybox-w32)

File: cli2gui/application/application.py [hunk 1]
```diff
@@ -3,22 +3,89 @@
...
+from .PySimpleGUI import Element, Window # local version
 
 from .. import c2gtypes
 from .pysimplegui2args import argFormat
 from .widgets import Widgets
 
+# TODO(milahu) better ...
+terminal_width = 80
+terminal_height = 24
+num_history_lines = 100 # aka "scrollback buffer size"
```
Comment (FredHappyface) [inline cli2gui/application/application.py:87]: Can you try and stick by the current style guide please? Not a problem at all for a draft but this needs fixing before a merge :)

File: cli2gui/application/application.py [hunk 2]
```diff
@@ -312,60 +392,160 @@ def createLayout(
...
-					widgets.stringSentencecase(
-						buildSpec["program_description"]
-						if buildSpec["program_description"]
-						else buildSpec["parser_description"]
-					)
+		],
+		[
+			widgets.label(
+				widgets.stringSentencecase(
+					buildSpec["program_description"] or
```
Comment (FredHappyface) [inline cli2gui/application/application.py:415]: Like what you've done with the following:\


```
buildSpec["program_description"]
						if buildSpec["program_description"]
						else buildSpec["parser_description"]

```

File: cli2gui/application/application.py [hunk 3]
```diff
@@ -312,60 +392,160 @@ def createLayout(
...
+		args_layout = addItemsAndGroups(section, args_layout, widgets) # magic here
 
 	# Set the layout
-	layout: list[list[Element]] = [[]]
-	if isinstance(menu, list):
-		layout: list[list[Element]] = [[pySimpleGui.Menu([["Menu", menu]], tearoff=True)]]
-
-	layout.extend(
+	# NOTE(milahu) one big array : )
+	layout: list[list[Element]] = \
```
Comment (FredHappyface) [inline cli2gui/application/application.py:402]: Can you explain the reason for changing this please? 

Also I think older versions of python crash with var: type = value even with importing annotations

File: cli2gui/application/application.py [hunk 4]
```diff
@@ -312,60 +392,160 @@ def createLayout(
...
+	log.info("cli2gui version 2022.1")
+	# TODO get version from pyproject.toml or from installed files
+
+	#log.info("log levels: " + "".join([l[0] for l in logging_levels_pretty]) + " = " + " ".join(logging_levels_pretty))
+	# TODO list only the active levels
+
+	#log.info("writing logfile: /path/to/cli2gui.log") # TODO
+
+	# logger test
+	if False:
```
Comment (FredHappyface) [inline cli2gui/application/application.py:511]: Guessing this is a temporary block of code?

File: cli2gui/application/application.py [hunk 5]
```diff
@@ -312,60 +392,160 @@ def createLayout(
...
+		'%(asctime)s %(levelname)-5s %(message)s' + "\n", # "-5s" -> align INFO and DEBUG/ERROR messages
+		# extra newline: make the log easier to read with wrapped lines
+		# time format:
+		#"%Y-%m-%d %H:%M:%S",
+		"%H:%M:%S",
+	)
+	[h.setFormatter(log_formatter) for h in log.handlers]
+
+	log.info("cli2gui version 2022.1")
+	# TODO get version from pyproject.toml or from installed files
```
Comment (FredHappyface) [inline cli2gui/application/application.py:503]: That'd be pretty awesome but happy enough with it hardcoded for now

File: cli2gui/application/application.py [hunk 6]
```diff
@@ -381,33 +561,547 @@ def run(buildSpec: c2gtypes.FullBuildSpec):
...
+	child_process_list_after_run = None
+	def child_process_watcher():
+
+		# FIXME avoid global, use queue or lock https://stackoverflow.com/a/19790994/10440128
+		global child_process_list, child_pid_set_before_run, child_process_list_after_run
+
+		proc = psutil.Process(pid=shell_process.pid)
+		# https://psutil.readthedocs.io/en/latest/#process-class
+		while True:
+			log.debug(f'child_process_watcher ---------------------------')
```
Comment (FredHappyface) [inline cli2gui/application/application.py:652]: Looks like an f-string isn't needed here. pylint should scream about this all being well

File: pyproject.toml [hunk 7]
```diff
@@ -27,12 +27,16 @@ readme = "README.md"
 [tool.poetry.dependencies]
 python = "^3.7"
 getostheme = "<2023,>=2021"
-Pillow = "<10,>=9.1.0"
-PySimpleGUI = "<5,>=4.59.0"
+#Pillow = "<10,>=9.1.0"
+Pillow = ">=9"
```
Comment (FredHappyface) [inline pyproject.toml:31]: Please try and use project tooling to manage this, setting Pillow = ">=9" breaks semver (even if not a perfect system)

File: pyproject.toml [hunk 8]
```diff
@@ -27,12 +27,16 @@ readme = "README.md"
...
+#PySimpleGUI = "<5,>=4.59.0"
+#PySimpleGUI: bundled locally. need unreleased git version for propagate=False
 PyYAML = "<7,>=6.0"
 catpandoc = {version = "<2024,>=2022.1", optional = true}
 PySimpleGUIWeb = {version = "<2,>=0.39.0", optional = true}
 PySimpleGUIQt = {version = "<2,>=0.35.0", optional = true}
+pyte = ">= 0.8.0"
```
Comment (FredHappyface) [inline pyproject.toml:38]: Should probably have these be optional if unix only packages

File: pyproject.toml [hunk 9]
```diff
@@ -27,12 +27,16 @@ readme = "README.md"
 [tool.poetry.dependencies]
 python = "^3.7"
 getostheme = "<2023,>=2021"
-Pillow = "<10,>=9.1.0"
-PySimpleGUI = "<5,>=4.59.0"
+#Pillow = "<10,>=9.1.0"
+Pillow = ">=9"
+#PySimpleGUI = "<5,>=4.59.0"
+#PySimpleGUI: bundled locally. need unreleased git version for propagate=False
```
Comment (FredHappyface) [inline pyproject.toml:33]: This needs to not be bundled locally before merging

**Review by FredHappyface — CHANGES_REQUESTED** [review]
Cheers for your work on this :). There's a few things that need looking at before this can be merged. Doesn't look very cross-platform which may be an issue

**Comment by milahu** [comment]
sorry for abandoning this

iirc, the big issues were

* to render ansi color escapes, we need to rewrite pyte-screen for tkinter.Text (bypass psg for better perf)
  * line wrapping in tkinter.Text is extremely slow, so we must do our own line wrapping (in pyte)
  * allow resizing the terminal = re-calculate all line-wraps
* bash job control (ctrl-z, jobs, bg, fg) was not working on windows (preexec_fn=os.setsid) (setsid is unix only)

for performance, its probably better to render the terminal with [alacritty](https://github.com/alacritty/alacritty/issues/450) or [kitty](https://github.com/kovidgoyal/kitty)
embedded into some widgets for menus and input forms

related

* https://github.com/lxqt/qtermwidget
* https://github.com/prompt-toolkit/ptterm
* https://github.com/spyder-ide/pywinpty
* https://github.com/Saadmairaj/tkterminal

im still surprised that qt/gtk/tk dont have cross-platform ansi terminal widgets ...

**Comment by FredHappyface** [comment]
Dw about it 

Yeah agreed that there don't appear to be tons of solutions. Though what confuses me about this all is why you wouldn't use use your fav terminal to start the process in the first place?

If you're done with this then I'll probably close the pr

As you probably know you are absolutely welcome to fork this if you want/ need this feature. Just for it to be part of this project it could really do with being platform independent

Thank you for all your time on this though. Definitely appreciated! - it's nice to see that the project is useful to others too

**Comment by milahu** [comment]
> why you wouldn't use use your fav terminal to start the process in the first place?

im thinking of "the user" who starts a gui app, enters his data via gui
and expects the process to run as if it was started from a terminal emulator
(but gui and terminal should be the same window)
as bonus, the user gets features like bash job control, or command history

anyway, closing as abandoned : /

**Pull request closed without merging**